Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h

```diff
@@ -0,0 +1,149 @@
+    };
+
+    // Will only return a MemoryMappedGPUBuffer, if gbm_bo allocation + mapping to userland + EGLImage creation succeeded.
+    static std::unique_ptr<MemoryMappedGPUBuffer> create(const IntSize&, OptionSet<BufferFlag>);
+
+    const IntSize& size() const { return m_size; }
+    const OptionSet<BufferFlag>& flags() const { return m_flags; }
+
+    // Map dma-buf into memory, if not yet mapped.
+    bool mapIfNeeded();
+    void unmapIfNeeded();
+
+    // Export gbm_bo buffer as dma-buf and wrap in EGLImage.
+    EGLImage createEGLImageFromDMABuf();
+
+    // Update the underlying data of the dma-buf, as often as desired.
+    // You need to obtain an AccessScope, fencing the write operation.
+    class AccessScope;
+    void updateContents(AccessScope&, const void* srcData, const IntRect& targetRect, unsigned bytesPerLine);
+    void updateContents(AccessScope&, const MemoryMappedGPUBuffer& srcBuffer, const IntRect& targetRect);
+
+    // You need to obtain an AccessScope, fencing the read operation.
+    std::span<uint32_t> mappedDataSpan(AccessScope&) const;
+
+    class AccessScope {
+        WTF_MAKE_NONCOPYABLE(AccessScope);
+        WTF_MAKE_FAST_ALLOCATED();
+    public:
+        ~AccessScope();
+
+        enum class Mode : bool {
+            Read,
+            Write
+        };
+
+        static std::unique_ptr<AccessScope> create(MemoryMappedGPUBuffer&, Mode);
+
+        const Mode& mode() const { return m_mode; }
+        const MemoryMappedGPUBuffer& buffer() const { return m_buffer; }
+
+    private:
+        AccessScope(MemoryMappedGPUBuffer&, Mode);
+
+        MemoryMappedGPUBuffer& m_buffer;
+        Mode m_mode { Mode::Read };
+    };
+
+    bool isMapped() const { return !!m_mappedData; }
+    bool isLinear() const;
+
+private:
+    MemoryMappedGPUBuffer(const IntSize&, OptionSet<BufferFlag>);
+
+    enum class DMABufSyncFlag : uint8_t {
+        Start = 1 << 0,
+        End   = 1 << 1,
+        Read  = 1 << 2,
+        Write = 1 << 3
+    };
+
+    bool performDMABufSyncSystemCall(OptionSet<DMABufSyncFlag> flags);
+    bool allocate(struct gbm_device*, const GLDisplay::DMABufFormat&);
+    bool createDMABufFromGBMBufferObject();
+    UnixFileDescriptor exportGBMBufferObjectAsDMABuf(unsigned planeIndex);
+
+    bool isLocked() const { return m_isLocked; }
+    void lock();
+    void unlock();
+
+    int primaryPlaneDmaBufFD() const;
+    uint32_t primaryPlaneDmaBufStride() const;
+
+    IntSize m_size;
+    OptionSet<BufferFlag> m_flags;
+    bool m_isLocked { false };
+    struct gbm_bo* m_bo { nullptr };
+    uint64_t m_modifier { 0 };
+    Vector<EGLAttrib> m_eglAttributes;
+    RefPtr<DMABufBuffer> m_dmaBuf;
+
+    void* m_mappedData { nullptr };
+    uint32_t m_mappedLength { 0 };
+};
+
+inline std::unique_ptr<MemoryMappedGPUBuffer::AccessScope> makeGPUBufferReadScope(MemoryMappedGPUBuffer& buffer)
+{
+    return MemoryMappedGPUBuffer::AccessScope::create(buffer, MemoryMappedGPUBuffer::AccessScope::Mode::Read);
+}
```

Comment:
Is this used? or do we plan to use this in follow up patches?

Reply:
If we want to toy with glReadPixels replacments, this would be needed, so I'd like to keep it.